File: code/C++/unit_tests/test_surrogate_generation.cpp
#ifndef SPISEME_HEADER
	#define SPISEME_HEADER
	#include "spiSeMe.h"
#endif
#include <fstream>
#include <sstream>

#define CATCH_CONFIG_MAIN
#include "catch.hpp"

TEST_CASE("Test surrogate generation functions"){

	// Load data
	std::ifstream input_file("test_data_A.dat", std::ifstream::in);
	double			temp_iei;
	std::string 		line;
	std::istringstream	linestream;
	std::vector<double>	data_A;
	while (std::getline(input_file, line)) {
		linestream.str(line);
		linestream >> temp_iei;
		data_A.push_back(temp_iei);
		linestream.clear();
	}
	input_file.close();

	std::vector< std::vector<double> > surrogates;
	std::vector<double>	data_Z(1, 0.0);
	spiSeMe_return_code	err;
	unsigned int M = 3;

	SECTION("Surrogate test JODI (arg)") {
		err = spiSeMe_surrogate_jodi(surrogates, data_Z, 1, true);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_jodi(surrogates, data_A, 0, true);
        	REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Surrogate test JODI (run)") {
		err = spiSeMe_surrogate_jodi(surrogates, data_A, M, false);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(surrogates.size() == M);
		double p, d;
		err = spiSeMe_distribution_test(p, d, data_A, surrogates[0]);
		REQUIRE(p == 1.0);
	}

	SECTION("Surrogate test IAAFT (arg)") {
		err = spiSeMe_surrogate_iaaft(surrogates, data_Z, "distribution", 1, true);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_iaaft(surrogates, data_A, "distribution", 0, true);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_iaaft(surrogates, data_A, "invalid", M, true);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Surrogate test IAAFT (run)") {
		err = spiSeMe_surrogate_iaaft(surrogates, data_A, "distribution", M, false);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(surrogates.size() == M);
		double p, d;
		err = spiSeMe_distribution_test(p, d, data_A, surrogates[0]);
		REQUIRE(p == 1.0);
	}

	SECTION("Surrogate test SA (arg)") {
		err = spiSeMe_surrogate_sa(surrogates, data_Z, 0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, -0.1, 1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, -1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 1.0, 0.1);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, 1.1);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, -1.1);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, 0.9, -1.0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, 0.9, 0.1, 1.0, "invalid");
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, 0.9, 0.1, 1.0, "max", 100, 50, 0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Surrogate test SA (run)") {
		err = spiSeMe_surrogate_sa(surrogates, data_A, 0.1, 1.0, 0.9, 0.1, 1.0, "max", 100, 50, M, false);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(surrogates.size() == M);
		double p, d;
		err = spiSeMe_distribution_test(p, d, data_A, surrogates[0]);
		REQUIRE(p == 1.0);
	}

	SECTION("Surrogate test Dither (arg)") {
		err = spiSeMe_surrogate_dither(surrogates, data_Z);
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_dither(surrogates, data_A, "invalid");
		REQUIRE(err == SSM_BAD_ARGUMENT);
		err = spiSeMe_surrogate_dither(surrogates, data_A, "uniform", 0.1, 0);
		REQUIRE(err == SSM_BAD_ARGUMENT);
	}

	SECTION("Surrogate test Dither (run)") {
		err = spiSeMe_surrogate_dither(surrogates, data_A, "uniform", 0.1, M, false);
		REQUIRE(err == SSM_SUCCESS);
		REQUIRE(surrogates.size() == M);
	}

}

File: code/Python/unit_tests/test_surrogate_generation.py
import unittest
import numpy as np
import sys
sys.path.insert(0, "../spiSeMe")
import spiSeMe_distribution_test
import spiSeMe_surrogate_dither
import spiSeMe_surrogate_iaaft
import spiSeMe_surrogate_jodi
import spiSeMe_surrogate_sa

class TestStringMethods(unittest.TestCase):
	def setUp(self):
		self.data_A = np.loadtxt('test_data_A.dat')
		self.data_Z = [1, 2, 3, 4, 5, 6]

	def test_surrogate_jodi_arg(self):
		print("JODI: Intentionally providing bad arguments...")
		self.assertFalse(spiSeMe_surrogate_jodi.spiSeMe_surrogate_jodi(self.data_Z))
		self.assertFalse(spiSeMe_surrogate_jodi.spiSeMe_surrogate_jodi(self.data_A, 'x'))
		self.assertFalse(spiSeMe_surrogate_jodi.spiSeMe_surrogate_jodi(self.data_A, 1, 'x'))
	def test_surrogate_jodi_run(self):
		print("JODI: Checking output formats...")
		self.M = 3
		self.data_out = spiSeMe_surrogate_jodi.spiSeMe_surrogate_jodi(self.data_A, self.M, False)
		self.assertTrue(isinstance(self.data_out,np.ndarray))
		self.assertEqual(self.data_out.ndim, 2)
		self.assertEqual(self.data_out.size, self.M * self.data_A.size)
		self.assertEqual(self.data_out[0].size, self.data_A.size)
		self.p, self.d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_out[0])
		self.assertEqual(self.p, 1)

	def test_surrogate_iaaft_arg(self):
		print("IAAFT: Intentionally providing bad arguments...")
		self.assertFalse(spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_Z))
		self.assertFalse(spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_A, 3.1))
		self.assertFalse(spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_A, 'invalid'))
		self.assertFalse(spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_A, 'distribution', 'x'))
		self.assertFalse(spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_A, 'distribution', 1, 'x'))
	def test_surrogate_iaaft_run(self):
		print("IAAFT: Checking output formats...")
		self.M = 3
		self.data_out = spiSeMe_surrogate_iaaft.spiSeMe_surrogate_iaaft(self.data_A, 'distribution', self.M, False)
		self.assertTrue(isinstance(self.data_out,np.ndarray))
		self.assertEqual(self.data_out.ndim, 2)
		self.assertEqual(self.data_out.size, self.M * self.data_A.size)
		self.assertEqual(self.data_out[0].size, self.data_A.size)
		self.p, self.d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_out[0])
		self.assertEqual(self.p, 1)

	def test_surrogate_sa_arg(self):
		print("SA: Intentionally providing bad arguments...")
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_Z, 0.1, 1.0))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 'x', 1.0))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 1.0, 0.1))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, -0.1, 1.0))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, -1.0))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 1.1))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, -1.0))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'invalid'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'max', 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'max', 100, 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'max', 100, 50, 'x'))
		self.assertFalse(spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'max', 100, 50, 1, 'x'))
	def test_surrogate_sa_run(self):
		print("SA: Checking output formats...")
		self.M = 3
		self.data_out = spiSeMe_surrogate_sa.spiSeMe_surrogate_sa(self.data_A, 0.1, 1.0, 0.9, 0.1, 1.0, 'max', 100, 50, self.M, False)
		self.assertTrue(isinstance(self.data_out,np.ndarray))
		self.assertEqual(self.data_out.ndim, 2)
		self.assertEqual(self.data_out.size, self.M * self.data_A.size)
		self.assertEqual(self.data_out[0].size, self.data_A.size)
		self.p, self.d = spiSeMe_distribution_test.spiSeMe_distribution_test(self.data_A, self.data_out[0])
		self.assertEqual(self.p, 1)

	def test_surrogate_dither_arg(self):
		print("Dither: Intentionally providing bad arguments...")
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_Z))
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 3.1))
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 'invalid'))
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 'uniform', 'x'))
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 'uniform', 0.1, 'x'))
		self.assertFalse(spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 'uniform', 0.1, 1, 'x'))
	def test_surrogate_dither_run(self):
		print("Dither: Checking output formats...")
		self.M = 3
		self.data_out = spiSeMe_surrogate_dither.spiSeMe_surrogate_dither(self.data_A, 'uniform', 0.01, self.M, False)
		self.assertTrue(isinstance(self.data_out,np.ndarray))
		self.assertEqual(self.data_out.ndim, 2)
		self.assertEqual(self.data_out.size, self.M * self.data_A.size)
		self.assertEqual(self.data_out[0].size, self.data_A.size)

if __name__ == '__main__':
    unittest.main()
